### PR TITLE
added example of how I think I wanna code the new permutation test

### PR DESCRIPTION
**Pull request by Souhatifour**
**What**
added example of how I think I wanna code the new permutation test

**Why**
As we are trying to assess how class imbalance impacts the predicted probabilities by comparing them to a baseline that reflects only the prior, it was important to control for the connectivity between nodes in the network (gene nodes degree)

**How**

1. Shuffling happens within bins
2. Positive labels are swapped with other positive-labeled candidates 
3. Negative labels are swapped similarly 

PR Checklist

- [ ]  Explained the purpose of this PR
- [ ]  Wrote example of how the code will look like
- [ ]  Updated documentation

### CODE REVIEW

File: src/example_structured_permutation_test.ipynb [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    "#2. bin the dataframe (in this example will just use 4)\n",
+    "\n",
+    "df['Bin']= pd.qcut(df['Degree'],4, labels=False)\n",
+    "\n",
+    "#3. function to replace a gene with another from the same bin & label\n",
+    "\n",
+    "def swap_gene(gene, bin,label,df):\n",
+    "    \n",
+    "    # gene must be swapped only by same bin and label genes\n",
+    "    same_bin_label_genes= df[(df['Bin']==bin)& (df['Label']== label)]['Gene'].tolist()\n",
```
Comment (phicks22) [inline src/example_structured_permutation_test.ipynb:87]: Consider a more concise variable name like `gene_group` for `same_bin_label_genes`

File: src/example_structured_permutation_test.ipynb [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    "    \n",
+    "    # here I ensure that the selected gene can't be replaced by itself\n",
+    "    \n",
+    "    same_bin_label_genes.remove(gene)\n",
+    "    \n",
+    "    # some bins might have only one gene, so we check for that here\n",
+    "    \n",
+    "    if same_bin_label_genes:\n",
+    "        \n",
+    "        return np.random.choice(same_bin_label_genes) # here we randomly choosing gene from same bin\n",
```
Comment (phicks22) [inline src/example_structured_permutation_test.ipynb:97]: Don't forget to set a seed for random operations for reproducibility. You can do this towards the top of the script/chunk

File: src/example_structured_permutation_test.ipynb [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    "    \n",
+    "    return gene # If no other genes in the bin, keep the original gene\n",
```
Comment (phicks22) [inline src/example_structured_permutation_test.ipynb:99]: Could returning the original gene interfere with the analysis?

File: src/example_structured_permutation_test.ipynb [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    "# here we applying the replacements\n",
+    "\n",
+    "df['Replaced_Gene'] = df.apply(lambda row: swap_gene(row['Gene'], row['Bin'], row['Label'], df), axis=1)\n",
+    "\n",
+    "print(df)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
```
Comment (phicks22) [inline src/example_structured_permutation_test.ipynb:110]: It doesn't seem like this chunk was executed before you submitted the PR. It's good to restart the kernel and run every cell in order before you submit a PR for notebooks.

File: src/example_structured_permutation_test.ipynb [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    "\n",
+    "df= pd.DataFrame({'Gene': genes, 'Degree':degrees, 'Label':labels})\n",
+    "\n",
+    "\n",
+    "print(df)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
```
Comment (phicks22) [inline src/example_structured_permutation_test.ipynb:52]: The execution count is out of order. Restart the kernel, execute each cell in order, then commit and push. You can also directly edit the `execution_count` key in the `.ipynb` metadata.

File: src/example_structured_permutation_test.ipynb [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+      "5  GeneF       8      1\n",
+      "6  GeneG      35     -1\n",
+      "7  GeneH      19      1\n",
+      "8  GeneI      50     -1\n",
+      "9  GeneJ      11     -1\n"
+     ]
+    }
+   ],
+   "source": [
+    "#toy example\n",
```
Comment (phicks22) [inline src/example_structured_permutation_test.ipynb:39]: Good work! This will easily turn into a test case when you write tests for this function.

**Review by phicks22 — CHANGES_REQUESTED** [review]
Nice! Short and sweet. Good documentation and easy to understand.

Last thing -> format the code with the `black` module. This is what I use. You can install it [here](https://github.com/psf/black) and see how to use it. You need to install a specific version for notebooks.

File: src/example_structured_permutation_test.ipynb (hunk 3 above)
Comment (Souhatifour) [reply src/example_structured_permutation_test.ipynb:99]: I was thinking to ask about that.

File: src/example_structured_permutation_test.ipynb (hunk 4 above)
Comment (Souhatifour) [reply src/example_structured_permutation_test.ipynb:110]: Correct, I coded it in a different notebook which included other analyses. Then, created a new notebook for this PR.

File: src/example_structured_permutation_test.ipynb (hunk 5 above)
Comment (Souhatifour) [reply src/example_structured_permutation_test.ipynb:52]: I do not see this in the original notebook.  This is funny 😄

File: src/example_structured_permutation_test.ipynb (hunk 2 above)
Comment (Souhatifour) [reply src/example_structured_permutation_test.ipynb:97]: Yep, I am adding that in the new PR. Nice catch!

File: src/example_structured_permutation_test.ipynb [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+   "cell_type": "code",
+   "execution_count": 34,
+   "id": "72ef4930",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/Users/souhatifour/anaconda3/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):\n",
```
Comment (phicks22) [inline src/example_structured_permutation_test.ipynb:63]: @Souhatifour Why are your models failing to converge?

File: src/example_structured_permutation_test.ipynb [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+    "\n",
+    "\n",
+    "########## here, I am trying to get the real predicted probs\n",
+    "\n",
+    "\n",
+    "# Train logistic regression on real data\n",
+    "X_real = np.vstack(df['embeddings'].values)\n",
+    "y_real = df['Label'] \n",
+    "\n",
+    "X_train_real, X_test_real, y_train_real, y_test_real = train_test_split(X_real, y_real, test_size=0.3, random_state=42)\n",
```
Comment (phicks22) [inline src/example_structured_permutation_test.ipynb:185]: @Souhatifour What does "real" mean in these variable names? Is it necessary?

File: src/example_structured_permutation_test.ipynb [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+    "\n",
+    "# train real model\n",
+    "clf_real = LogisticRegression(max_iter=100)\n",
```
Comment (phicks22) [inline src/example_structured_permutation_test.ipynb:188]: @Souhatifour Do you know why you're using Logistic regression with full L2 regularization? In Txt2Onto models we use ElasticNet with `C=1.0` and `l1_ratio=0.4`. It would be good to replicate this analysis using the hyperparameters that the networks people use.

Comment (phicks22) [inline src/example_structured_permutation_test.ipynb:188]: @Souhatifour What does `clf` mean?

File: src/example_structured_permutation_test.ipynb [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+    "    df['Replaced_Embeddings']= df['Replaced_Gene'].map(embeddings)\n",
+    "\n",
+    "    # here, we are turnning them into a feature matrix using np.vstack\n",
+    "\n",
+    "    X= np.vstack(df['Replaced_Embeddings'].values) \n",
+    "    y= df['Gene']\n",
+    "\n",
+    "    # split data for 70% training and 30% testing\n",
+    "\n",
+    "    X_train, X_test, y_train, y_test= train_test_split(X,y,test_size=0.3,random_state=42)\n",
```
Comment (phicks22) [inline src/example_structured_permutation_test.ipynb:242]: @Souhatifour It's good to set a constant variable at the top of the script/notebook for random seed/states. So at the top have something like `SEED=42` and just use the `SEED` variable wherever you need to throughout the script. Otherwise, you might end up with different random seeds/states throughout the notebook if you ever changed one.

File: src/example_structured_permutation_test.ipynb [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+    "    })\n",
+    "\n",
+    "    results.append(test_results)\n",
+    "\n",
+    " # combine results from all runs\n",
+    "final_results = pd.concat(results, ignore_index=True)\n",
+    "\n",
+    "\n",
+    "# combine shuffled results\n",
+    "final_results = pd.concat(results, ignore_index=True)\n",
```
Comment (phicks22) [inline src/example_structured_permutation_test.ipynb:264]: @Souhatifour Duplicated `final_results` assignment.

**Comment by phicks22** [comment]
@Souhatifour What is a real vs predicted probability?

**Review by phicks22 — COMMENTED** [review]
@Souhatifour Please see my comments.

File: src/example_structured_permutation_test.ipynb (hunk 1 above)
Comment (phicks22) [reply src/example_structured_permutation_test.ipynb]: Thanks for changing.
